### DISCUSSION TITLE
dynamic version

### DISCUSSION THREAD

**Discussion by surgiie** [bashly-framework/bashly#290] [Ideas]
Is there a way bashly could support a dynamic version number? It would be useful if the version number could optionally be defined as a environment variable in `initialize.sh` or some place equivalent. This would allow authors to dynamically set the version based on a `git tag` for example.

**Comment by surgiie** [comment]
Actually looks like i can put a command to be executed under `version` in the yaml already.

**Reply by DannyBen** [reply]
> Actually looks like i can put a command to be executed under version in the yaml already.

Where did you see this?

It would help if you show some concrete example.
Do you want it to be dynamic and generated when the user runs the script? Or dynamic as in get the git sha of **your** current branch or something?

The `--version` command just echos the `$version` variable, which is set in the `initialize()` function.
If you want to override it, the only way I can think of now is to edit the `src/initialize.sh` file and add `version=anything` there.

**Reply by surgiie** [reply]
It was a shot in the dark to see if bashly supported dynamic values directly in the yaml, so I tried putting a git command in the `bashly.yaml` `version field and this appears to work:

```yaml
# bashly.yml
version: $(cd $(dirname $(which my-cli)) && git describe --abbrev=0 && cd ~-)
```

Not sure this is intended but it works.

I can also try the `$version` variable as an alternative if you believe that is more appropriate.

**Reply by DannyBen** [reply]
What you did is fine. If you look at the generated script, in the `initialize()` function, you will see that it just placed your string in it as:

```
version="$(...)"
```

I am not sure this is what you intend to do - is it?
This will be dynamic **at run time**, not at build time - this command will run for anyone who runs your script.

If this is your intention, this option is ok.
If your intention is to have this dynamic value evaluate at build time, then it is not good.

For build time dynamic version update, you will have too cook something up yourself. Bashly does not have such a feature.

One option would be to run something like this (⚠️ careful: this will modify your bashly.yml):

```bash
# replace "version: ANYTHING" with "version: GIT-SHA"
sed -Ei "s/version: .*/version: $(git rev-parse --short HEAD)/" src/bashly.yml
```

If this approach works for you, you can create a `build.sh` script in the root directory, which will do this followed by a `bashly generate`.
Personally, I am using [opcode](https://github.com/DannyBen/opcode) for things like this. Then it would be something like this:

```yaml
# op.conf
build: op update-version && op generate
generate: bashly generate
update-version: sed -Ei "s/version: .*/version: $(git rev-parse --short HEAD)/" src/bashly.yml
```

then I run `op build`

**Reply by surgiie** [reply]
I think im okay with what ive got for now, run time version seems okay.
